esp-hal/src/radio_clock_ctrl.rs

```diff
@@ -0,0 +1,171 @@
+        #[cfg(wifi)]
+        {
+            any_clocks_enabled |= self.wifi_clock_enabled.load(Ordering::Relaxed);
+        }
+        #[cfg(ieee802154)]
+        {
+            any_clocks_enabled |= self.ieee802154_clock_enabled.load(Ordering::Relaxed);
+        }
+
+        any_clocks_enabled
+    }
+    /// Enable or disable the clock.
+    fn set_clock_status(&self, enabled: bool, radio_peripheral: RadioPeripherals) {
+        critical_section::with(|cs| {
+            let mut radio_clock = self.radio_clock.borrow_ref_mut(cs);
+            if enabled {
+                radio_clock.enable(radio_peripheral)
+            } else {
+                radio_clock.disable(radio_peripheral)
+            }
+        })
+    }
+    /// Enable or disable the PHY clock.
+    ///
+    /// If the PHY clock state was successfully changed, or the current state
+    /// already matched the specified one `Ok(())` is returned.
+    /// If the PHY clock is being disabled, but other modem clocks are still
+    /// active `Err(ModemClocksEnabledError)` is returned.
+    fn set_phy_clock_status(&self, enabled: bool) -> Result<(), ModemClocksEnabledError> {
+        // We shouldn't disable the PHY clock, if other clocks are still active.
+        if !enabled && self.any_modem_clocks_enabled() {
+            return Err(ModemClocksEnabledError);
+        }
+        // If the current clock status is already the same as the user provided one, we
+        // return. Otherwise we store the new value.
+        if PHY_CLOCK_ENABLED
+            .compare_exchange(!enabled, enabled, Ordering::Relaxed, Ordering::Relaxed)
+            .is_err()
+        {
+            return Ok(());
+        }
+        self.set_clock_status(enabled, RadioPeripherals::Phy);
+        Ok(())
+    }
+    /// Enable or disable a modem clock.
+    ///
+    /// If required, this will also enabled or disable the PHY clock.
+    /// # Panics
+    /// This panics, if [RadioPeripherals::Phy] is passed in, since the PHY
+    /// clock is controlled internally.
+    pub fn set_modem_clock_status(&self, enabled: bool, radio_peripheral: RadioPeripherals) {
+        // We first assert, that the clock we're touching isn't the PHY clock, since
+        // that's special.
+        assert_ne!(radio_peripheral, RadioPeripherals::Phy);
+
+        // We obtain a reference to the AtomicBool tracking the state of the clock.
+        let modem_clock_enabled = match radio_peripheral {
+            #[cfg(bt)]
+            RadioPeripherals::Bt => &self.bt_clock_enabled,
+            #[cfg(wifi)]
+            RadioPeripherals::Wifi => &self.wifi_clock_enabled,
+            #[cfg(ieee802154)]
+            RadioPeripherals::Ieee802154 => &self.ieee802154_clock_enabled,
+            _ => unreachable!(),
+        };
+        // If the states already match, we return. Otherwise the new state is stored.
+        if modem_clock_enabled
+            .compare_exchange(!enabled, !enabled, Ordering::Relaxed, Ordering::Relaxed)
```

Comment:
```suggestion
            .compare_exchange(!enabled, enabled, Ordering::Relaxed, Ordering::Relaxed)
```

Reply:
Oh I overlooked that. Thanks!

Reply:
This seems to have reappeared accidentally :)

Reply:
This should be fixed now. Hopefully